### PR TITLE
Add support for bpf2bpf function calls

### LINKED ISSUE #451: Prevail does not support function calls
Hello, I'm writing a packet parser xdp program which only has an `xdp` section. The program passes verification and loads successfully with the verifier in the linux kernel, I'm trying to get it verified with PREVAIL as well, since windows is a possible target.

While trying to run it through the verifier with:
`sudo ./ebpf-verifier/check myxdp.o xdp`
The verifier returns an error:
`error: Can't find any maps sections in file myxdp.o`

The program was compiled with the command:
`clang -O2 -g -Wall -target bpf -c myxdp.c -o myxdp.o`

llvm-objdump output shows the section headers as follows:
```
linux@linux-pc:~/dev$ llvm-objdump --section-headers myxdp.o

myxdp.o:	file format elf64-bpf

Sections:
Idx Name                   Size     VMA              Type
  0                        00000000 0000000000000000 
  1 .strtab                000002c0 0000000000000000 
  2 .text                  00000f60 0000000000000000 TEXT
  3 .rel.text              00000040 0000000000000000 
  4 xdp                    000008c0 0000000000000000 TEXT
  5 .relxdp                00000010 0000000000000000 
  6 .rodata.str1.1         000000b0 0000000000000000 DATA
  7 license                00000004 0000000000000000 DATA
  8 .debug_loclists        00000662 0000000000000000 DEBUG
  9 .debug_abbrev          000002bd 0000000000000000 DEBUG
 10 .debug_info            00000bda 0000000000000000 DEBUG
 11 .rel.debug_info        00000060 0000000000000000 
 12 .debug_rnglists        000001ea 0000000000000000 DEBUG
 13 .debug_str_offsets     00000294 0000000000000000 DEBUG
 14 .rel.debug_str_offsets 00000a30 0000000000000000 
 15 .debug_str             00000668 0000000000000000 DEBUG
 16 .debug_addr            00000130 0000000000000000 DEBUG
 17 .rel.debug_addr        00000250 0000000000000000 
 18 .BTF                   00001668 0000000000000000 
 19 .rel.BTF               00000010 0000000000000000 
 20 .BTF.ext               000015f0 0000000000000000 
 21 .rel.BTF.ext           000015f0 0000000000000000 
 22 .debug_frame           000000e8 0000000000000000 DEBUG
 23 .rel.debug_frame       00000120 0000000000000000 
 24 .debug_line            000007ef 0000000000000000 DEBUG
 25 .rel.debug_line        00000150 0000000000000000 
 26 .debug_line_str        000000cf 0000000000000000 DEBUG
 27 .llvm_addrsig          00000002 0000000000000000 
 28 .symtab                000005d0 0000000000000000
```

A marginally simpler program verifies fine with the tool. The examples provided in the repo work as well, so I can confirm the tool is set up correctly.

Any help would be appreciated, Thank you.

### PR DESCRIPTION
**Pull request by dthaler**
Implemented approach discussed in meeting between Dave, Elazar, Arie, and others.
Each function is treated as a "macro" and CFG nodes are created as if the compiler had inlined the functions.  The new nodes use variables prefixed with the conceptual call stack, so that nesting and multiple calls to the same function are both supported.  E.g., "1/5/r6.svalue=0" means the svalue property of r6 in a macro called from label 5 which was in turn called from label 1.

As tested by the bpf_conformance suite, it turns out that an Exit instruction is responsible for restoring registers R6-R9 (as opposed to being restored by other instructions generated by the compiler).  Hence the CallLocal instruction makes copies of R6-R9 state, and the Exit instruction restores R6-R9 from the copies, and cleans up the copied state. 

This PR also:
* fixes a use-after-free bug in wto.cpp where args2 state was being read after a pop() that could free that memory.
* adds a "simplify" option for YAML tests that want to test behavior without no_simplify. Previously all YAML tests used no_simplify, but we needed to test CallLocal both with and without the simplify step in order to fully test the CFG changes.
* updates the template matching code in test_marshal.cpp to correctly deal with fields with wildcards, to resolve an issue hit when testing this PR.

Fixes #451

### CODE REVIEW

**Comment by elazarg** [comment]
I don't think handling function calls as macros at the CFG level is the right approach, and I do not believe Arie and @caballa supported this approach. Inlining (as opposed to computing a summary) can be implemented by letting the abstract interpreter treat the code in same way a concrete interpreter would:  jump to the start of the function with a copy of the abstract state, execute normally, then jump back with the resulting state.

In order to have the same behavior as in the macro approach you'd need to index the tracked states (instead of the instructions) with the same conceptual call stack, but I think this approach is more flexible in case we'd want to implement alternative approaches to inlining. For example, we can join on entry from all caller sites and thus converge faster, probably with a simpler debug output (also maybe slightly more useful proof-carrying-code in the future).

**Comment by caballa** [comment]
I read the slides given by @dthaler and asked Arie about the meeting you had and what I understand @dthaler has implemented is consistent with what I remember. 

@elazarg  suggest a "dynamic" inlining approach. By "dynamic", I mean that the abstract interpreter actually does the inlining as it goes. There are several advantages of doing that. Elazar mentioned some. Another important advantage is that a function call wouldn't be inlined if it's unreachable. 

What @dthaler has implemented is more like a "static" inlining approach by doing some program transformation (I haven't looked at the PR but this is my guess). 

Both approaches are compatible. We can start with Dave's approach since it's already implemented and tested and we can always implement what Elazar suggests if needed.

**Comment by dthaler** [comment]
Code coverage failure seems to be a transient failure due to coveralls backend, not specific to this PR.

File: src/asm_syntax.hpp [hunk 1]
```diff
@@ -11,24 +11,28 @@
 #include "crab/variable.hpp"
 #include "spec_type_descriptors.hpp"
 
+constexpr char STACK_FRAME_DELIMITER = '/';
+
 namespace crab {
 struct label_t {
```
Comment (elazarg) [inline src/asm_syntax.hpp:17]: My personal taste: rename label to something else ("_label_t" maybe) and add a new type named "label_t" that holds the label and the context.
Also I'd prefer the context to have an easier to handle type (`vector<_label_t>`).

File: src/crab/ebpf_domain.cpp [hunk 2]
```diff
@@ -1541,7 +1541,21 @@ void ebpf_domain_t::operator()(const Un& stmt) {
...
+        return;
+    for (int r = R6; r <= R9; r++) {
+        for (data_kind_t kind = data_kind_t::types; kind <= data_kind_t::stack_numeric_sizes;
+             kind = (data_kind_t)((int)kind + 1)) {
+            variable_t src_var = variable_t::stack_frame_var(kind, r, prefix);
+            if (!m_inv[src_var].is_top())
+                assign(variable_t::reg(kind, r), src_var);
+            havoc(src_var);
+        }
+    }
```
Comment (elazarg) [inline src/crab/ebpf_domain.cpp:1557]: Maybe add methods `save_callee_saved_registers()` and `restore_callee_saved_registers()` near `scratch_caller_saved_registers()`.

File: src/crab/ebpf_domain.cpp [hunk 3]
```diff
@@ -1723,6 +1737,16 @@ crab::interval_t ebpf_domain_t::get_map_max_entries(const Reg& map_fd_reg) const
     return result;
 }
 
+void ebpf_domain_t::operator()(const ValidCall& s) {
+    if (!s.stack_frame_prefix.empty()) {
+        EbpfHelperPrototype proto = global_program_info->platform->get_helper_prototype(s.func);
+        if (proto.return_type == EBPF_RETURN_TYPE_INTEGER_OR_NO_RETURN_IF_SUCCEED) {
```
Comment (elazarg) [inline src/crab/ebpf_domain.cpp:1743]: Is this information available "statically" before the analysis?

**Review by elazarg — COMMENTED** [review]
I'm sorry about the delay. I still have my reservations, but this approach does the job.

Below I added some comments and proposed additional tests. The most important comment is probably the handling of correlations in callee-saved registers. Let me know if you think it should be addressed now or later on.

File: src/crab/cfg.hpp [hunk 4]
```diff
@@ -315,6 +315,8 @@ class cfg_t final {
 
     [[nodiscard]] size_t size() const { return static_cast<size_t>(std::distance(begin(), end())); }
 
+    void dump_cfg(std::string str);
+
```
[the remark this answers is not in the document]
Comment (dthaler) [reply src/crab/cfg.hpp]: Done

File: src/crab/ebpf_domain.cpp (hunk 2 above)
Comment (dthaler) [reply src/crab/ebpf_domain.cpp]: Done

**Comment by shankarseal** [comment]
@dthaler @elazarg the [eBPF for Windows](http://github.com/ebpf-for-windows) project would like to onboard bpf2bpf function call support to investigate if this can replace tail calls (among others) to improve performance of bpf programs in network datapath. Can you please continue this PR (which seems stalled) and resolve the remaining issues?

File: src/crab/ebpf_domain.cpp [hunk 5]
```diff
@@ -1541,7 +1541,21 @@ void ebpf_domain_t::operator()(const Un& stmt) {
...
+void ebpf_domain_t::operator()(const Exit& a) {
+    // Clean up any state for the current stack frame.
+    std::string prefix = a.stack_frame_prefix;
+    if (prefix.empty())
+        return;
+    for (int r = R6; r <= R9; r++) {
+        for (data_kind_t kind = data_kind_t::types; kind <= data_kind_t::stack_numeric_sizes;
+             kind = (data_kind_t)((int)kind + 1)) {
+            variable_t src_var = variable_t::stack_frame_var(kind, r, prefix);
+            if (!m_inv[src_var].is_top())
```
[the remark this answers is not in the document]
Comment (elazarg) [reply src/crab/ebpf_domain.cpp]: Added commented-out YAML test.

**Comment by elazarg** [comment]
@shankarseal done.

**Pull request merged**